### DISCUSSION TITLE
Connect to DLNA/UPnP server and/or client?

### DISCUSSION THREAD

**Discussion by Slesa** [strawberrymusicplayer/strawberry#1631] [Ideas]
Are there already efforts to support DLNA/UPnP servers? I have a Raspberry Pi with a minidlna running, so it would a perfect match to connect both fruits.
I also have a Sonos system runnin, and I think they support the same protocol. So maybe they can be chosen as output devices, too.

What I have found so far is the vlc-qt library, so it should not be that hard to implement. Or maybe Platinum/Neptun can be an option. 

But anyway, did someone else already start working on this?

**Comment by reezer** [comment]
Something like a built-in DLNA/UPnP server to make the collection available would be amazing.

**Comment by Slesa** [comment]
To act as a server seems to be not that complicated. All it seems to need is to reply to a udp broadcast message and deliver a XML file with all needed information.
I am currently trying out the other option and got connection to my speakers and the minidlna server with simple Qt methods. I am still struggling to understand the description fields, e.g. how to detect the speaker / server.

**Reply by Slesa** [reply]
Ok, it is a bit more. I think there are few steps, or better blocks:
- Identify Media Servers by broadcast. Implement the SOAP commands to get its content
- Identify Media Renderer by broadcast. Implement the SOAP commands to send a playlist.
  Bonus: implement volume control
- Implement the reply to the broadcast to tell we are a Media Server, plus the replies for the incoming SOAP command to send a playlist
- Implement the reply to the broadcast to tell we are a Media Renderer, plus the replies for the incoming SOAP command to receive a playlist
  Bonus: implement volume control
- Support grouping of rooms

But when done, we automatically have a remote control - each other UPnp controller.

**Comment by Slesa** [comment]
Via the library QtUPnP I could easily read the files from a media server and play them locally. They are provided via http link.
I am wondering if those remote files should be registered in the database, because the browsing is uncomfortable and annoying. Any thoughts?
